### DISCUSSION TITLE
Should DateTimeInterface be obsolete as well?

### DISCUSSION THREAD

**Discussion by ghost** [symfony/symfony#52390] [Q&A]
I'm opening a discussion rather than issue despite of fact I know the response from the Core Team will be little (it's mostly stof that answers discussions) 

I think it's unfortunate that at the moment Symfony uses DateTimeInterface as TH because this means we still support mutable version of DateTime (interface can't be implemented in the userland code so it's only DateTime vs DateTimeImmutable). We did a great job removing mutable version from the codebase but I think we should have gone one step further and discourage using mutable version by devs by forcing them to use immutable version if they want to use Symfony.

(if you add thumb down please justify in the comments section)

**Comment by ro0NL** [comment]
the DateTimeInterface itself is immutable. AFAIK Symfony wont anticipate DateTime anymore in its codebase.

> discourage using mutable version by devs

im not sure it's a job for Symfony, ideally PHP solved this with simple `DateTime::withTimezone()` et al (while deprecating the setters).

**Reply by ghost** [reply]
> the DateTimeInterface itself is immutable

Do you mean this interface is immutable by design or it's immutable because by coincidence all the methods it has do not care whether a concrete implementation is mutable or immutable?

> im not sure it's a job for Symfony

I know the conversation what's framework's job is hard. However it always comes down to the specific problem, in this case DateTime vs DateTimeImmutable. Mutable DateTime is flawed and it should have never been introduced (wrong decision back in PHP 5.5 where the PHP core did not have the courage to revert a bad decision introduced in 5.2). In my opinion the lead tools and their maintainers should have the courage to make hard and what's more the **right** decisions because who else will be? Some random John Doe from China? If Symfony switched to the DateTimeImmutable it would be a huge signal for the PHP world, hopefully leading to the the removal of mutable DateTime from the core.

**Reply by stof** [reply]
Symfony always instantiate DateTimeImmutable now (except for cases where it is explicitly requested to provide a DateTime object).

But this does not mean that Symfony cannot be more flexible in the kind of input it accepts.

**Reply by ghost** [reply]
@stof this is what I want to point out - I know frameworks should be flexible but they should also show a good practices, be a role model. Right now when we have DateTimeInterface TH we show that we are fine with DateTime. I think Symfony could go one step further and show DateTime is wrong by not accepting it as a method argument.
